sockeye/decoder.py

```diff
@@ -274,48 +299,30 @@ def _step(self,
-        attention_input = self.attention.make_input(seq_idx, word_vec_prev, rnn_output)
+        attention_input = self.attention.make_input(seq_idx, word_vec_prev, attention_rnn_output)
         attention_state = attention_func(attention_input, attention_state)
 
-        # (3) Combine context with hidden state
-        if self.context_gating:
-            # context: (batch_size, encoder_num_hidden)
-            # gate: (batch_size, rnn_num_hidden)
-            gate = mx.sym.FullyConnected(data=mx.sym.concat(word_vec_prev, rnn_output, attention_state.context, dim=1),
-                                         num_hidden=self.num_hidden, weight=self.gate_w, bias=self.gate_b)
-            gate = mx.sym.Activation(data=gate, act_type="sigmoid",
-                                     name="%sgate_activation_t%d" % (self.prefix, seq_idx))
-
-            # mapped_rnn_output: (batch_size, rnn_num_hidden)
-            mapped_rnn_output = mx.sym.FullyConnected(data=rnn_output,
-                                                      num_hidden=self.num_hidden,
-                                                      weight=self.mapped_rnn_output_w,
-                                                      bias=self.mapped_rnn_output_b,
-                                                      name="%smapped_rnn_output_fc_t%d" % (self.prefix, seq_idx))
-            # mapped_context: (batch_size, rnn_num_hidden)
-            mapped_context = mx.sym.FullyConnected(data=attention_state.context,
-                                                   num_hidden=self.num_hidden,
-                                                   weight=self.mapped_context_w,
-                                                   bias=self.mapped_context_b,
-                                                   name="%smapped_context_fc_t%d" % (self.prefix, seq_idx))
-
-            # hidden: (batch_size, rnn_num_hidden)
-            hidden = mx.sym.Activation(data=gate * mapped_rnn_output + (1 - gate) * mapped_context,
-                                       act_type="tanh",
-                                       name="%snext_hidden_t%d" % (self.prefix, seq_idx))
-
-        else:
-            # hidden: (batch_size, rnn_num_hidden)
-            hidden = mx.sym.FullyConnected(data=mx.sym.concat(rnn_output, attention_state.context, dim=1),
+        # (3) Upper layers if present or combine with attention
+        if self.upper_layers_rnn:
+            # TODO: context gating for multiple layers (with attention)
+            upper_rnn_output, upper_rnn_layer_states = \
+                self.upper_layers_rnn(attention_rnn_output, attention_state.context,
+                                      state.layer_states[self.attention_rnn_n_states:])
+            # TODO: if we do not include a last layer here, inference complains
+            #       about duplicate variable output names (??). Have to investigate
+            #       more.
+            # Do we want to include attention another time here?
```

Comment:
This part is the one where I still have to tweak a little bit. If we leave attention in, this is effectively the same as below if no context gating is used, so it can be simplified.